### PR TITLE
Support double precision in cuda::polarToCart() + fix broken tests

### PR DESCRIPTION
**Pull request by cv3d**
Merge with opencv/opencv_extra#522
related: #12561, #12268

Fix for `cuda::polarToCart` functionality
### This pullrequest changes
- Support double precision
- test double precision and tune tolerance

All tests are performed using
```
opencv_perf_cudaarithm --gtest_filter=*PolarToCart* --perf_min_samples=10000 --perf_force_samples=10000 
```

|       | Type  | mean | median | min  |    stddev   | total time | --gtest_param_filter     |
|-------|-------|------|--------|------|-------------|------------|--------------------------|
| Before| float | 0.20 |  0.19  | 0.18 | 0.01 (5.3%) |  2054 ms   | (1920x1080, true)        |
|       | double| N/A  |  N/A   |  N/A |  N/A        |   N/A      |  N/A                     |
| After | float | 0.19 |  0.19  | 0.18 | 0.01 (4.1%) |  1994 ms   | (1920x1080, 32FC1, true) |
|       | double| 0.45 |  0.45  | 0.39 | 0.02 (4.2%) |  4587 ms   | (1920x1080, 64FC1, true) |

```
allow_multiple_commits=1
force_builders=Custom
docker_image:Custom=ubuntu-cuda:16.04
```

### CODE REVIEW

File: modules/cudaarithm/src/cuda/polar_cart.cu [hunk 1]
```diff
@@ -157,20 +157,20 @@ void cv::cuda::cartToPolar(InputArray _x, InputArray _y, OutputArray _mag, Outpu
...
 
-        const float mag_val = useMag ? mag(y, x) : 1.0f;
-        const float angle_val = angle(y, x);
+        const T mag_val = useMag ? mag(y, x) : 1.0f;
+        const T angle_val = angle(y, x);
 
-        float sin_a, cos_a;
-        ::sincosf(scale * angle_val, &sin_a, &cos_a);
+        T sin_a, cos_a;
+        ::sincos(scale * angle_val, &sin_a, &cos_a);
```
Comment (alalek) [inline modules/cudaarithm/src/cuda/polar_cart.cu]: What is about performance regressions?

> sincosf
> sincos

The first one is for "float" input, second for "double" input. [CUDA reference](https://docs.nvidia.com/cuda/cuda-math-api/group__CUDA__MATH__SINGLE.html#group__CUDA__MATH__SINGLE)

File: modules/core/include/opencv2/core/cvdef.h [hunk 2]
```diff
@@ -205,9 +205,9 @@ enum CpuFeatures {
 
 
 /* fundamental constants */
-#define CV_PI   3.1415926535897932384626433832795
-#define CV_2PI  6.283185307179586476925286766559
-#define CV_LOG2 0.69314718055994530941723212145818
+#define CV_PI   3.14159265358979323846264338327950288419716939937510
+#define CV_2PI  6.28318530717958647692528676655900576839433879875021
+#define CV_LOG2 0.69314718055994530941723212145817656807550013436026
```
Comment (alalek) [inline modules/core/include/opencv2/core/cvdef.h]: Such global changes must be done as a separate patch.

**Comment by cv3d** [comment]
> What is about performance regressions?

I recovered the original code. Is that enough, or could you please emphasize a little bit?

**Comment by alalek** [comment]
Need to run performance tests with and without patch to ensure that changes don't introduce performance regressions.
Because CUDA here is for performance. A lot of decisions in CUDA modules between accuracy and performance conflicts are done to improve the performance at first.

**Comment by cv3d** [comment]
> The first one is for "float" input, second for "double" input.

Actually, `sincosf` internally calls `sincos`, so there is no performance difference
```C++
__func__(void sincosf(float a, float *sptr, float *cptr))
{
  double s, c;

  sincos((double)a, &s, &c);
  *sptr = (float)s;
  *cptr = (float)c;
}
```

**Comment by alalek** [comment]
... in what CUDA SDK version?

CUDA 8.0 SDK has this:
```
__DEVICE_FUNCTIONS_STATIC_DECL__ void __sincosf(float a, float *sptr, float *cptr) __THROW
{
  __nv_fast_sincosf(a, sptr, cptr);
}
```

**Comment by cv3d** [comment]
That is strange!
Mine is also NVIDIA GPU Computing Toolkit  - CUDA v8.0
And the function is defined in `math_functions.hpp` lines 1722-1729
Is that because mine is Windows version?

**Comment by alalek** [comment]
It is `__sincosf` (with double underscores) - approximated version.

But there is another one for `sincosf` (math_functions.hpp, line 1252):
```
__MATH_FUNCTIONS_DECL__ void sincosf(float a, float *sptr, float *cptr) __THROW
{
  if (__USE_FAST_MATH__) {
    __nv_fast_sincosf(a, sptr, cptr);
  } else {
    __nv_sincosf(a, sptr, cptr);
  }
}
```
and this 'double' version (line 1716):
```
__func__(void sincos(double a, double *sptr, double *cptr))
{
  *sptr = sin(a);
  *cptr = cos(a);
}
```

P.S. Only performance tests can answer on performance regressions. For best result it is necessary to test changes on different devices (platforms): from old (like GTX650 - supported yet?), to modern (the same platform as 1080). For minimal insurance of the result lets test changes on some device.

**Comment by cv3d** [comment]
@alalek I updated the description with regression results.

File: modules/cudaarithm/perf/perf_element_operations.cpp [hunk 3]
```diff
@@ -1423,17 +1424,19 @@ PERF_TEST_P(Sz_AngleInDegrees, CartToPolar,
 //////////////////////////////////////////////////////////////////////
 // PolarToCart
 
-PERF_TEST_P(Sz_AngleInDegrees, PolarToCart,
+PERF_TEST_P(Sz_Type_AngleInDegrees, PolarToCart,
             Combine(CUDA_TYPICAL_MAT_SIZES,
+                    testing::Values(CV_32FC1, CV_64FC1),
                     Bool()))
```
Comment (alalek) [inline modules/cudaarithm/perf/perf_element_operations.cpp]: There is `CUDA_SANITY_CHECK()` below.
It is very sensitive to test name: run perf test binary with "--perf_verify_sanity".
So, need to update sanity data too (opencv_exta/testdata/perf) or "disable" these checks (like `SANITY_CHECK_NOTHING()`).

Comment (cv3d) [reply modules/cudaarithm/perf/perf_element_operations.cpp]: I made a PR against extras repo.
BTW: I noticed that the single-precision version is not very consistent with its results. This have nothing to do with proposed PR though. It seems we might need to relax the perf tolerance for it as well.

**Comment by alalek** [comment]
Please rebase to 3.4 branch.
I will test my "merge-3.4" scripts with CUDA updates on this patch.

**Comment by cv3d** [comment]
@alalek Done!

**Comment by alalek** [comment]
Performance is without regressions on 1080.

Please update [documentation](https://github.com/opencv/opencv/blob/3.4.3/modules/cudaarithm/include/opencv2/cudaarithm.hpp#L416-L425) too.
There is corresponding `cartToPolar` function. It will be good to keep API consistent for similar functions (can be done in another PR).

**Comment by cv3d** [comment]
> Performance is without regressions on 1080
Could you please emphasize a little bit?

**Comment by alalek** [comment]
I mean, no performance regressions are observed (no problems are here).

**Comment by cv3d** [comment]
I see! I also have 1080. Perhaps I just made a bogus report!
Sometime I find myself testing against outdated build 😖

**Comment by cv3d** [comment]
@alalek A kind reminder!
